### PR TITLE
task 7

### PR DESCRIPTION
**Pull request by Isaac-Madden**

### CODE REVIEW

**Comment by Isaac-Madden** [comment]
Sorry I got my naming conventions messed up and forgot to switch branches. 

This branch now includes answers for "8-**patch**-api-articles" as well as "7-**post**-api-articles"

Hope thats okay!

File: __tests__/app.test.js [hunk 1]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+    .send( {username: "icellusedkars", body: "also a test"} )
+    .then( data => {
+      expect(data.body.comment_added).toEqual(
+        expect.objectContaining({
+          comment_id: 19,
+          body: "also a test",
+          article_id: 5,
+          author: "icellusedkars",
+          votes: 0,
+          created_at: expect.any(String),
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:230]: Really good test

File: __tests__/app.test.js [hunk 2]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+  })
+
+  test("error passed to handlers, responds with 404 when username violates foreign key constraint", () => {
+
+    return request(app)
+     .post("/api/articles/3/comments")
+     .send( { username: "isaac", body: "this shouldn't work" } )
+     .expect(404)
+     .then( data => {
+      expect(data.body.msg).toBe("not found");
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:243]: Great test

File: __tests__/app.test.js [hunk 3]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+  test("responds with 400 and error if body has not been defined", () => {
+
+    return request(app)
+    .post("/api/articles/5/comments")
+    .send( {username: "icellusedkars"} )
+    .expect(400)
+    .then( data => {
+      expect(data.body.msg).toBe("incomplete request, please add a comment")
+   })
+  })
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:279]: Testing for the POST is brilliant, one optional test you can add is that it still completes the post even if there are additional unnecessary keys in the request body

File: __tests__/app.test.js [hunk 4]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+}) // end of "POST /api/articles/:article_id/comments" testing
+
+describe("PATCH: /api/articles/:article_id", () => {
+
+  test("returns 202 status code following successful request", () => {
+
+   return request(app)
+    .patch("/api/articles/5")
+    .send( { inc_votes: 35 } )
+    .expect(202)
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:290]: Nice try with the `202` code, but as it's defined as "the request has been accepted for processing, but the processing has not been completed" I don't think it's applicable here, it's just a boring old `200`

File: __tests__/app.test.js [hunk 5]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+      .then( data => {
+        expect(data.body.article).toMatchObject({
+          article_id: 9,
+          title: "They're not exactly dogs, are they?",
+          topic: 'mitch',
+          author: 'butter_bridge',
+          body: 'Well? Think about it.',
+          created_at: "2020-06-06T09:10:00.000Z",
+          votes: -5,
+          article_img_url: 'https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700'
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:324]: Very nice test

File: __tests__/app.test.js [hunk 6]
```diff
@@ -202,4 +202,148 @@ describe("api/articles/:article_id/comments", () => {
...
+      })
+  })
+
+  test("returns 400 and error when no votes object passed", () => {
+    return request(app)
+      .patch("/api/articles/9")
+      .send()
+      .expect(400)
+      .then( data => {
+        expect(data.body.msg).toBe("no votes submitted")
```
Comment (Pipe-Smoking-Rabbit) [inline __tests__/app.test.js:335]: You could make this a `200` and respond with an unchanged article. We might want to add the ability to PATCH more than just the votes, if the `inc_votes` key is missing we might not want that to always be an error. So by making this a `200` our testing is open to expansion

File: Server/controllers/articles.controller.js [hunk 7]
```diff
@@ -19,4 +19,20 @@ exports.getAllArticles = (req, res, next) => {
...
+   
+    if (inc_votes === undefined) {
+        return next({ status: 400, msg: "no votes submitted" })
+    }
+
+    updateArticleByID(article_id, inc_votes)
+    .then( article => {
+        res.status(202).send({ article })
+    })
+    .catch( err => next(err) )
```
Comment (Pipe-Smoking-Rabbit) [inline Server/controllers/articles.controller.js:37]: If you like there is a shorthand syntax for the catch 
`.catch(next)` which will invoke next with the error

File: Server/controllers/comments.controller.js [hunk 8]
```diff
@@ -9,4 +9,25 @@ exports.getCommentsByArticleID = (req, res, next) => {
...
+    const userName = req.body.username
+    const commentBody = req.body.body
+
+    if (userName === undefined) {
+        return next({ status: 400, msg: "incomplete request, provide username" })
+    }
+
+    if(commentBody === undefined) {
+        return next({ status: 400, msg: "incomplete request, please add a comment" }) 
+    }
```
Comment (Pipe-Smoking-Rabbit) [inline Server/controllers/comments.controller.js:26]: These two conditions are not necessary, if you are using the `$` binding syntax in your query then PSQL will throw an error for you if either of these values are undefined

**Review by Pipe-Smoking-Rabbit — APPROVED** [review]
All looks very good, there's a few comments on the test suite and a comment for a bit of unnecessary logic in the POST controller but otherwise very clean code! ✅

**Pull request merged**